dali/pipeline/basic/tensor.h

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_BASIC_TENSOR_H_
+#define DALI_PIPELINE_BASIC_TENSOR_H_
+
+#include "dali/common.h"
+#include "dali/pipeline/data/tensor.h"
+
+namespace dali {
+namespace basic {
+
+template <size_t N>
+struct TensorShape {
+  explicit TensorShape(const Tensor<CPUBackend>& t) {
+    for (size_t i = 0; i < N; i++) {
+      shape[i] = t.dim(i);
+    }
+  }
+
+  explicit TensorShape(const TensorShape& other) {
+    for (size_t i = 0; i < N; i++) {
+      shape[i] = other.shape[i];
+    }
+  }
+
+  // Why call for TensorWrapper (subclass) calls this instead of proper copy constructor??
+  template <typename Iter>
+  explicit TensorShape(Iter it) {
+    for (size_t i = 0; i < N; i++) {
+      shape[i] = *it;
+      ++it;
+    }
+  }
+
+  const std::array<Index, N>& GetShape() const { return shape; }
+
+ private:
+  std::array<Index, N> shape;
+};
+
+template <typename T, size_t N>
+struct TensorWrapper : public TensorShape<N> {
```

Comment:
I wanted to set part of the interface hare and avoid having to expose it by some accessors if I put shape as member. But you're right, it will be better if TesnsorShape is a separate thing and to use composition.

Reply:
A note on naming (quite important in a public API, though): TensorWrapper sounds like something wrapped around a tensor. TensorView would be more adequate, even though it suffers the same shortcoming, just to a smaller extent.

Reply:
Removed from this PR.